compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp

```diff
@@ -34,79 +34,105 @@
-            connectionOp.getTarget().getDefiningOp()) == op) {
-      // Handle producer connection operations.
-      FailureOr<AMDAIE::NpuCircularDmaCpyNdOp> maybeNpuDmaUserOp =
-          connectionOp.getNpuCircularDmaCpyNdUser();
-      if (failed(maybeNpuDmaUserOp)) {
-        return connectionOp.emitOpError()
-               << "does not have a circular DMA op user";
-      }
-      FailureOr<size_t> maybeNewRepetitionCount = getUpdatedRepetitionCount(
-          maybeNpuDmaUserOp->getTargetMixedSizes(),
-          maybeNpuDmaUserOp->getTargetMixedStrides(), repetitionCount);
-      if (failed(maybeNewRepetitionCount)) {
-        return maybeNpuDmaUserOp->emitOpError() << "no repetition count found";
-      }
-      repetitionCount = maybeNewRepetitionCount.value();
-    } else if (auto connectionOp = dyn_cast<AMDAIE::ConnectionOp>(userOp);
-               connectionOp.getSource() &&
-               dyn_cast_if_present<LogicalObjFifoOpInterface>(
-                   connectionOp.getSource().getDefiningOp()) == op) {
-      // Handle consumer connection operations.
+    if (auto connectionOp = dyn_cast<AMDAIE::ConnectionOp>(userOp)) {
       FailureOr<AMDAIE::NpuCircularDmaCpyNdOp> maybeNpuDmaUserOp =
           connectionOp.getNpuCircularDmaCpyNdUser();
-      if (failed(maybeNpuDmaUserOp)) {
-        return connectionOp.emitOpError()
-               << "does not have a circular DMA op user";
+
+      if (failed(maybeNpuDmaUserOp)) continue;
+
+      AMDAIE::NpuCircularDmaCpyNdOp npuDma = maybeNpuDmaUserOp.value();
+
+      if (connectionOp.getTarget() &&
+          dyn_cast_if_present<LogicalObjFifoOpInterface>(
+              connectionOp.getTarget().getDefiningOp()) == op) {
+        appendRepetitionCount(npuDma.getTargetMixedSizes(),
+                              npuDma.getTargetMixedStrides());
       }
-      FailureOr<size_t> maybeNewRepetitionCount = getUpdatedRepetitionCount(
-          maybeNpuDmaUserOp->getSourceMixedSizes(),
-          maybeNpuDmaUserOp->getSourceMixedStrides(), repetitionCount);
-      if (failed(maybeNewRepetitionCount)) {
-        return maybeNpuDmaUserOp->emitOpError() << "no repetition count found";
+
+      if (connectionOp.getSource() &&
+          dyn_cast_if_present<LogicalObjFifoOpInterface>(
+              connectionOp.getSource().getDefiningOp()) == op) {
+        appendRepetitionCount(npuDma.getSourceMixedSizes(),
+                              npuDma.getSourceMixedStrides());
       }
-      repetitionCount = maybeNewRepetitionCount.value();
     }
   }
-  return repetitionCount;
+
+  // merge the repetition counts:
+  if (repetitionCounts.empty()) return 1;
+  size_t combinedRepetitionCount = 1;
+
+  // Sort and unique-ify the repetition counts:
+  std::sort(repetitionCounts.begin(), repetitionCounts.end());
+  auto last = std::unique(repetitionCounts.begin(), repetitionCounts.end());
+  repetitionCounts.erase(last, repetitionCounts.end());
```

Comment:
Why is the sort + uniquify needed? Can't you just look for the smallest `repetitionCount`?

Reply:
Not needed, I've removed it.